prov/verbs_ud/src/verbs_ud_ep.c

```diff
@@ -0,0 +1,807 @@
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "verbs_ud.h"
+
+static int ofi_verbs_ud_ep_close(fid_t ep_fid);
+static int ofi_verbs_ud_ep_control(fid_t ep_fid,
+				   int command,
+				   void *arg);
+static int ofi_verbs_ud_ep_bind(fid_t ep_fid,
+				struct fid *bfid,
+				uint64_t flags);
+static int ofi_verbs_ud_ep_setname(fid_t ep_fid, void *addr,
+				   size_t addrlen);
+static int ofi_verbs_ud_ep_getname(fid_t ep_fid, void *addr,
+				   size_t *addrlen);
+
+static struct fi_ops ofi_verbs_ud_fi_ops = {
+	.size = sizeof(ofi_verbs_ud_fi_ops),
+	.close = ofi_verbs_ud_ep_close,
+	.bind = ofi_verbs_ud_ep_bind,
+	.control = ofi_verbs_ud_ep_control,
+	.ops_open = fi_no_ops_open,
+};
+
+static struct fi_ops_cm ofi_verbs_ud_cm_ops = {
+	.size = sizeof(ofi_verbs_ud_cm_ops),
+	.setname = ofi_verbs_ud_ep_setname,
+	.getname = ofi_verbs_ud_ep_getname,
+	.getpeer = fi_no_getpeer,
+	.connect = fi_no_connect,
+	.listen = fi_no_listen,
+	.accept = fi_no_accept,
+	.reject = fi_no_reject,
+	.shutdown = fi_no_shutdown,
+};
+
+static struct fi_ops_ep ofi_verbs_ud_ep_ops = {
+	.size = sizeof(ofi_verbs_ud_ep_ops),
+	.cancel = fi_no_cancel,
+	.getopt = fi_no_getopt,
+	.setopt = fi_no_setopt,
+	.tx_ctx = fi_no_tx_ctx,
+	.rx_ctx = fi_no_rx_ctx,
+	.rx_size_left = fi_no_rx_size_left,
+	.tx_size_left = fi_no_tx_size_left,
+};
+
+static inline
+int ofi_verbs_ud_cq_comp(struct util_cq *util_cq, struct ibv_wc *wc)
+{
+	struct fi_cq_tagged_entry *comp;
+	int ret = FI_SUCCESS;
+	struct ofi_verbs_ud_wr_entry *wr_entry;
+
+	fastlock_acquire(&util_cq->cq_lock);
+	if (unlikely(ofi_cirque_isfull(util_cq->cirq))) {
```

Comment:
We should protect against queue overruns by controlling how many operations are active on the associated tx/rx queues, or simply avoid reaping completions from the underlying cq's if these are full.